examples/stream-in-distribute-out/contracts/base/StreamInDistributeOut.sol

```diff
@@ -0,0 +1,294 @@
+} from "@superfluid-finance/ethereum-contracts/contracts/apps/SuperAppBase.sol";
+import {
+    IInstantDistributionAgreementV1,
+    IDAv1Library
+} from "@superfluid-finance/ethereum-contracts/contracts/apps/IDAv1Library.sol";
+
+// //////////////////////////////////////////////////////////////
+// ERRORS
+// //////////////////////////////////////////////////////////////
+
+/// @dev Thrown when the wrong token is streamed to the contract.
+error InvalidToken();
+
+/// @dev Thrown when the `msg.sender` of the app callbacks is not the Superfluid host.
+error Unauthorized();
+
+/// @title Abstract contract to stream into and distribute out.
+/// @notice Users stream in and receive a proportional amount of shares. The shares represent a
+/// percentage of a distribution, which gets called in the `executeAction` function.
+/// @dev Inheriting contracts MUST implement `_beforeDistribution()` in inheriting contracts.
+abstract contract StreamInDistributeOut is SuperAppBase {
+
+    // //////////////////////////////////////////////////////////////
+    // EVENTS
+    // //////////////////////////////////////////////////////////////
+
+    /// @dev Emits when action is successfully executed.
+    /// @param distributionAmount Amount that gets distributed to the index.
+    event ActionExecuted(uint256 distributionAmount);
+    
+    /// @dev Emits when action fails in a stream termination callback AND the `amountOwed` can NOT
+    /// be transferred back to the address closing the stream.
+    /// @param amountOwed Amount owed back to the address that closed the stream.
+    event ActionFailed(uint256 amountOwed);
+
+    // //////////////////////////////////////////////////////////////
+    // VARIABLES
+    // //////////////////////////////////////////////////////////////
+
+    /// @dev Last Distribution timestamp. Used to compute the amount owed to an address that closes
+    /// a stream but the `executeAction` call fails.
+    uint256 public lastDistribution;
+
+    /// @dev IDAv1Library for brevity.
+    using IDAv1Library for IDAv1Library.InitData;
+    IDAv1Library.InitData internal _idaLib;
+
+    /// @dev Superfluid Contracts.
+    ISuperfluid internal immutable _host;
+    IConstantFlowAgreementV1 internal immutable _cfa;
+
+    /// @dev SuperToken to stream in.
+    ISuperToken internal immutable _inToken;
+
+    /// @dev SuperToken to distribute out.
+    ISuperToken internal immutable _outToken;
+
+    /// @dev Index ID for the distribution.
+    uint32 internal constant INDEX_ID = 0;
+
+    // //////////////////////////////////////////////////////////////
+    // MODIFIERS
+    // //////////////////////////////////////////////////////////////
+
+    /// @dev Checks every callback to validate inputs. MUST be called by the host.
+    /// @param token The Super Token streamed in. MUST be the in-token.
+    modifier validCallback(ISuperToken token) {
+        if (token != _inToken && token != _outToken) revert InvalidToken();
```

Comment:
```
require(_isCFAv1(agreementClass), "LotterySuperApp: only CFAv1 supported");
```

should always check like this, please check our existing example code base too

Reply:
After checking the rest of examples, we are very inconsistent about this very important check. Opening a new issue to keep track.

Reply:
This can not revert, because `approveSubscription` triggers either the `afterAgreementCreated` or `afterAgreementUdpated` (depending on if the subscription exists).
https://github.com/superfluid-finance/protocol-monorepo/blob/089010c1403a930c392435ec173dc37c76689cfc/packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol#L327

This prevents anyone from approving subscriptions if the callbacks revert when the agreement is the IDA.

---

We are checking at the start of every callback and only returning the `ctx` if it's not the CFA.